ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -968,6 +995,73 @@ private void applyTtlUpdate(MessageInfo messageInfo, RemoteReplicaInfo remoteRep
     }
   }
 
+  /**
+   * Applies an undelete to the blob described by {@code messageInfo}.
+   * @param messageInfo the {@link MessageInfo} that will be transformed into a TTL update
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   * @throws StoreException
+   */
+  private void applyUndelete(MessageInfo messageInfo, RemoteReplicaInfo remoteReplicaInfo) throws StoreException {
+    DataNodeId remoteNode = remoteReplicaInfo.getReplicaId().getDataNodeId();
+    try {
+      messageInfo = new MessageInfo(messageInfo.getStoreKey(), messageInfo.getSize(), messageInfo.isDeleted(),
+          messageInfo.isTtlUpdated(), true, messageInfo.getExpirationTimeInMs(), messageInfo.getCrc(),
+          messageInfo.getAccountId(), messageInfo.getContainerId(), messageInfo.getOperationTimeMs(),
+          messageInfo.getLifeVersion());
+      remoteReplicaInfo.getLocalStore().undelete(messageInfo);
+      logger.trace("Remote node: {} Thread name: {} Remote replica: {} Key undelete id: {}", remoteNode, threadName,
+          remoteReplicaInfo.getReplicaId(), messageInfo.getStoreKey());
+    } catch (StoreException e) {
+      // The blob may be undeleted, which is alright
+      if (e.getErrorCode() == StoreErrorCodes.Life_Version_Conflict
+          || e.getErrorCode() == StoreErrorCodes.ID_Undeleted) {
+        logger.trace("Remote node: {} Thread name: {} Remote replica: {} Key already undeleted: {}", remoteNode,
+            threadName, remoteReplicaInfo.getReplicaId(), messageInfo.getStoreKey());
+      } else {
+        throw e;
+      }
+    }
+    // A Repair event for an undelete signifies that an undelete message was received from the remote and it is fired
+    // as long as the undelete is guaranteed to have taken effect locally.
+    if (notification != null) {
+      notification.onBlobReplicaUndeleted(dataNodeId.getHostname(), dataNodeId.getPort(),
+          messageInfo.getStoreKey().getID(), BlobReplicaSourceType.REPAIRED);
+    }
+  }
+
+  /**
+   * Applies a delete to the blob described by {@code messageInfo}.
+   * @param messageInfo the {@link MessageInfo} that will be transformed into a TTL update
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   * @throws StoreException
+   */
+  private void applyDelete(MessageInfo messageInfo, RemoteReplicaInfo remoteReplicaInfo) throws StoreException {
+    DataNodeId remoteNode = remoteReplicaInfo.getReplicaId().getDataNodeId();
+    try {
+      messageInfo =
+          new MessageInfo(messageInfo.getStoreKey(), messageInfo.getSize(), false, messageInfo.isTtlUpdated(), false,
```

Comment:
Should the deleted flag be set to true here?

Reply:
good catch. Although those flags are not used in the delete method, this would definitely cause confusion. Thanks for pointing out.